nep-5.mediawiki

```diff
@@ -19,55 +19,91 @@ As the NEO blockchain scales, Smart Contract deployment and invocation will beco
+
+This method MUST always return the same value every time it is invoked.
 
 ====balanceOf====
 
-* Syntax: <code>public static BigInteger balanceOf(byte[] account)</code>
+<pre>
+public static BigInteger balanceOf(byte[] account)
+</pre>
 
-* Remarks: "balanceOf" returns the token balance of the '''account'''.
+Returns the token balance of the <code>account</code>.
+
+The parameter <code>account</code> SHOULD be a 20-byte address. If not, this method SHOULD <code>throw</code> an exception.
+
+If the <code>account</code> is an unused address, this method SHOULD return <code>0</code>.
 
 ====transfer====
 
-* Syntax: <code>public static bool transfer(byte[] from, byte[] to, BigInteger amount)</code>
+<pre>
+public static bool transfer(byte[] from, byte[] to, BigInteger amount)
+</pre>
+
+Transfers an <code>amount</code> of tokens from the <code>from</code> account to the <code>to</code> account.
+
+The parameters <code>from</code> and <code>to</code> SHOULD be 20-byte addresses. If not, this method SHOULD <code>throw</code> an exception.
+
+The parameter <code>amount</code> SHOULD be greater than or equal to <code>0</code>. If not, this method SHOULD <code>throw</code> an exception.
 
-* Remarks: "transfer" will transfer an '''amount''' of tokens from the '''from''' account to the '''to''' account. Contract '''MUST''' check the <code>payable</code> flag of the receiver contracts to decide whether it should transfer the tokens to those contracts.
+The function SHOULD return <code>false</code> if the <code>from</code> account balance does not have enough tokens to spend.
+
+If the method succeeds, it MUST fire the <code>transfer</code> event, and MUST return <code>true</code>, even if the <code>amount</code> is <code>0</code>, or <code>from</code> and <code>to</code> are the same address.
+
+The function SHOULD check whether the <code>from</code> address equals the caller contract hash. If so, the transfer SHOULD be processed; If not, the function SHOULD use the SYSCALL <code>Neo.Runtime.CheckWitness</code> to verify the transfer.
+
+If the <code>to</code> address is a deployed contract, the function SHOULD check the <code>payable</code> flag of this contract to decide whether it should transfer the tokens to this contract.
+
+If the transfer is not processed, the function SHOULD return <code>false</code>.
 
 ===Events===
 
 ====transfer====
 
-* Syntax: <code>public static event Action<byte[], byte[], BigInteger> transfer</code>
+<pre>
+public static event transfer(byte[] from, byte[] to, BigInteger amount)
+</pre>
+
+MUST trigger when tokens are transferred, including zero value transfers.
 
-* Remarks: The "transfer" event is raised after a successful execution of the "transfer" method.
+A token contract which creates new tokens SHOULD trigger a <code>transfer</code> event with the <code>from</code> address set to <code>null</code> when tokens are created.
```

Comment:
+1000 on this!

This was not in the original NEP5, but block explorers like neotracker or neoscan depend on this event to be called when tokens are minted. So there have been cases where this notification was not emitted at the time of minting, and resulted in confusion for users who would see their tokens in their wallets, but not on the explorer. Until they make a transfer.

Hopefully this will prevent that from happening in the future.

That being said, would it be possible to update this from a `SHOULD` to a `MUST`?

Reply:
After further consideration, I am starting to wonder if it makes more sense to remove this clause from the `transfer` event. The ask is to have the `transfer` event triggered from a method that is not a part of the NEP5 standard. Whether it be though a token sale, direct allocation method, or some dynamic token creation event, these all result in an action more akin to that of a `mintToken` or `mint` event, rather than a `transfer` that is inferred to be a `mint` when the `from"`address is `null`.

The desire for a separate event for token creation (and even token burn) have been expressed a few times in other posts.

https://github.com/neo-project/proposals/pull/39

https://github.com/CityOfZion/neo-python/issues/457

I would like to reaffirm the proposal stated in the latter thread, with more urgency on the `mint` or `mintToken` event over the `burn` event.

#### mint or mintToken
```
public static event mint(byte[] to, BigInteger amount)
public static event mintToken(byte[] to, BigInteger amount)
```
MUST trigger when tokens are minted.


#### burn or burnToken
```
public static event burn(byte[] from, BigInteger amount)
public static event burnToken(byte[] from, BigInteger amount)
```
MUST trigger when tokens are burned.

These events will help to increase transparency into contract mint and burn events amongst users, and possibly improve confidence that tokens are not being minted unknowingly.

Reply:
Too many events. The standard should be as simple as possible.

Reply:
While I agree that the standard should be simple to understand and implement, shouldn't it also be descriptively accurate?

Rather than adding the SHOULD clause to the `transfer` event, it seems more accurate to create a separate event called `mint`. I understand that the implication that tokens are being transferred from `null` because they are being created, but this implication assumes this prior knowledge from a reader rather than being clear and self explanatory.

Could we please reconsider just the addition of the `mint` event to clean up the notifications being emitted by contracts?

Reply:
I don't quite see the need for a mint/burn event myself either. Seeing funds created/destroyed by coming from/to null shows that the tokens either came from nowhere or were sent to nowhere. 

+1 on changing the SHOULD to a MUST though. If tokens are being created/destroyed it is good to be able to pick this up without checking TotalSupply constantly and inferring the change. Having the change notification arise out of a certain block is a much cleaner way to show it.

Reply:
https://www.ietf.org/rfc/rfc2119.txt